src/context/index.ts

```diff
@@ -0,0 +1,232 @@
+	})
+	languages.registerCodeActionsProvider('*', {
+		provideCodeActions: (doc, _range, context) => {
+			if (context.only) return
+			const result = context.diagnostics[0]?._result
+			panel.select(result)
+
+			const editor = window.visibleTextEditors.find(editor => editor.document === doc)
+			if (!editor) return // When would editor be undef?
+			if (!result) return // Don't clear the decorations until the next result is selected.
+
+			const locations = result?.codeFlows?.[0]?.threadFlows?.[0]?.locations ?? []
+			const messages = locations.map((tfl, i) => {
+				const text = tfl.location?.message?.text
+				return `Step ${i + 1}${text ? `: ${text}` : ''}`
+			})
+			const ranges = locations.map(tfl => regionToSelection(doc, parseRegion(tfl.location?.physicalLocation?.region)))
+			const rangesEnd = ranges.map(range => {
+				const endPos = doc.lineAt(range.end.line).range.end
+				return new Range(endPos, endPos)
+			})
+			const rangesEndAdj = rangesEnd.map(range => {
+				const tabCount = doc.lineAt(range.end.line).text.match(/\t/g)?.length ?? 0
+				const tabCharAdj = tabCount * (editor.options.tabSize as number - 1) // Intra-character tabs are counted wrong.
+				return range.end.character + tabCharAdj
+			})
+			const maxRangeEnd = Math.max(...rangesEndAdj) + 2 // + for Padding
+			const decorCallouts = rangesEnd.map((range, i) => ({
+				range,
+				hoverMessage: messages[i],
+				renderOptions: { after: { contentText: ` ${'┄'.repeat(maxRangeEnd - rangesEndAdj[i])} ${messages[i]}`, } }, // ←
+			}))
+			editor.setDecorations(decorationTypeCallout, decorCallouts)
+			editor.setDecorations(decorationTypeHighlight, ranges)
+			return []
+		}
+	})
+
+	// Virtual Documents
+	workspace.registerTextDocumentContentProvider('sarif', {
+		provideTextDocumentContent: (uri, token) => {
+			const [logUriEncoded, runIndex, artifactIndex] = uri.path.split('/')
+			const logUri = decodeURIComponent(logUriEncoded)
+			const artifact = store.logs.find(log => log._uri === logUri)?.runs[+runIndex]?.artifacts?.[+artifactIndex]
+			const contents = artifact?.contents
+			if (contents?.text) return contents?.text
+			if (contents?.binary) {
+				const lines = Buffer.from(contents?.binary, 'base64').toString('hex').match(/.{1,32}/g) ?? []
+				return lines.reduce((sum, line, i) => {
+					const lineNo = ((i + 128) * 16).toString(16).toUpperCase().padStart(8, '0')
+					const preview = Buffer.from(line, 'hex').toString('utf8').replace(/(\x09|\x0A|\x0B|\x0C|\x0D|\x1B)/g, '?')
+					return `${sum}${lineNo}  ${line.toUpperCase().match(/.{1,2}/g)?.join(' ')}  ${preview}\n`
+				}, '')
+			}
+			token.isCancellationRequested = true
+			return ''
+		}
+	})
+
+	// API
+	return {
+		async openLogs(logs: Uri[]) {
+			store.logs.push(...await loadLogs(logs))
+			if (store.results.length) panel.show()
+		},
+		async closeLogs(logs: Uri[]) {
+			for (const uri of logs) {
+				store.logs.removeWhere(log => log._uri === uri.toString())
+			}
+		},
```

Comment:
This should remove all diagnostics, decorations, code-lens, etc. as well.

Reply:
Same for class all logs.

---
In reply to: [434672138](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r434672138) [](ancestors = 434672138)

Reply:
Diagnostics removals are handled here: [`observe(store.logs, () => workspace.textDocuments.forEach(setDiags))`](https://github.com/microsoft/sarif-vscode-extension/blob/1d83faaaf100aca58b834afe69c3fe471875d104/src/context/index.ts#L68)

Decorations have been fixed and are now handled [here](https://github.com/microsoft/sarif-vscode-extension/blob/1d83faaaf100aca58b834afe69c3fe471875d104/src/context/index.ts#L133)

We currently don't have any code lenses. I think that covers all types.